2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +153,245 @@ Delegate(sender, validator, amount) ==
     /\ lastTx' = [ id |-> nextTxId, tag |-> "delegate", fail |-> fail,
                    sender |-> sender, toAddr |-> validator, value |-> amount ]
     /\ failed' = (fail \/ failed)
+    /\ nextTxId' = nextTxId + 1
+    /\ txCounter' = txCounter + 1
+    /\ UNCHANGED <<epoch, unbonded, idUnbonds, slashes, idSlashes, enqueuedSlashes, frozenValidators>>
     /\  IF fail
         THEN
-          UNCHANGED <<balanceOf, unbonded, delegated, bonded>>
+          UNCHANGED <<balanceOf, posAccount, bonded, idBonds, totalDeltas>>
         ELSE
           \* transaction succeeds
           \* update the balance of 'sender'
           /\ balanceOf' = [ balanceOf EXCEPT ![sender] = @ - amount]
-          /\ delegated' = [ n \in 1..UnbondingLength, user \in UserAddrs, val \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender /\ val = validator
-                     THEN delegated[n, user, val] + amount
-                     ELSE delegated[n, user, val]
-                    ]
-          /\ bonded' = [ n \in 1..UnbondingLength, user \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender
-                     THEN bonded[n, user] + amount
-                     ELSE bonded[n, user]
-                    ]
-          /\ UNCHANGED unbonded
-
-
-\* unbond `amount` tokens from a validator
+          /\ posAccount' = posAccount + amount
+          /\ bonded' = [ bonded EXCEPT ![sender, validator] = @ \union {[ id |-> idBonds,
+                                                                          epoch |-> epoch + PipelineLength,
+                                                                          amount |-> amount ]}]
+          /\ idBonds' = idBonds + 1
+          \* updates totalDeltas from PipelineLength to UnbondingLength 
+          /\ totalDeltas' = [ n \in 0..2*UnbondingLength, val \in ValidatorAddrs |-> totalDeltas[n, val] + 
+                              IF n >= UnbondingLength + PipelineLength /\ val = validator
```

Comment:
`n >= UnbondingLength + PipelineLength` can you explain this condition?

Reply:
index unbonding_length of total_deltas is cur_epoch. Then we want to update total deltas starting from cur_epoch+pipeline_lenght = unbonding_length+pipeline_lenght

Reply:
Wouldn't it be clearer to use `epoch` here (which is `cur_epoch` IIUC) instead of `UnbondingLength`?